File: adventofcode/2022/day05/day05.rb
# frozen_string_literal: true

def read_input
  ARGF.readlines
end

def parse_stacks(lines, count)
  stacks = []
  count.times do |i|
    stack = ""
    lines.each do |line|
      index = i * 4 + 1
      stack = line[index] + stack unless line[index]&.empty?
    end
    stacks.push(stack.strip)
  end
  stacks
end

def move(stacks, count, from, to)
  fstr = stacks[from]
  stacks[to] += fstr[-count..].reverse
  stacks[from] = fstr[0...(fstr.size - count)]
end

def apply_moves(lines, stacks)
  lines.each do |line|
    if /move (\d+) from (\d+) to (\d+)/ =~ line
      move(stacks, $1.to_i, $2.to_i - 1, $3.to_i - 1)
    end
  end
end

def part1(stacks, moves)
  apply_moves(moves, stacks)
  stacks.map {|s| s[-1]}.join
end

def move2(stacks, count, from, to)
  fstr = stacks[from]
  stacks[to] += fstr[-count..]
  stacks[from] = fstr[0...(fstr.size - count)]
end

def apply_moves2(lines, stacks)
  lines.each do |line|
    if /move (\d+) from (\d+) to (\d+)/ =~ line
      move2(stacks, $1.to_i, $2.to_i - 1, $3.to_i - 1)
    end
  end
end

def part2(stacks, moves)
  apply_moves2(moves, stacks)
  stacks.map {|s| s[-1]}.join
end

lines = read_input
sindex = lines.find_index { |line| line.strip.start_with?("1") }

scount = lines[sindex].split.map(&:strip).size
stacks = parse_stacks(lines[0...sindex], scount)

puts part1 stacks.dup, lines[sindex..]
puts part2 stacks.dup, lines[sindex..]

File: adventofcode/2022/day05/day05.go
package main

import (
	"bufio"
  "regexp"
	"fmt"
	"os"
  "strings"
  "strconv"
)

func readInput() []string {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

type Move struct {
  count int;
  from int;
  to int;
}

func reverse(s string) string {
  runes := []rune(s)
  for i, j := 0, len(runes)-1; i < j; i, j = i + 1, j - 1 {
    runes[i], runes[j] = runes[j], runes[i]
  }
  return string(runes)
}

func (m *Move) apply(stacks []string) {
  fstr := stacks[m.from]
  stacks[m.to] = stacks[m.to] + reverse(fstr[len(fstr)-m.count:])
  stacks[m.from] = fstr[0:len(fstr)-m.count]
}

func (m *Move) apply2(stacks []string) {
  fstr := stacks[m.from]
  stacks[m.to] = stacks[m.to] + fstr[len(fstr)-m.count:]
  stacks[m.from] = fstr[0:len(fstr)-m.count]
}

func parseStacks(lines []string) ([]string, []Move) {
  sindex := 0
  for index, line := range(lines) {
    if strings.HasPrefix(strings.TrimSpace(line), "1") {
      sindex = index
      break
    }
  }
  scount := len(strings.Fields(lines[sindex]))

  stacks := []string{}
  for i := 0; i < scount; i++ {
    stack := ""
    for _, line := range(lines[0:sindex]) {
      index := i * 4 + 1
      stack = fmt.Sprintf("%c%s", line[index], stack)
    }
    stacks = append(stacks, strings.TrimSpace(stack))
  }

  moves := []Move{}
  r := regexp.MustCompile(`move (\d+) from (\d+) to (\d+)`)
  for _, line := range(lines[sindex+2:]) {
    res := r.FindAllStringSubmatch(line, -1)[0]
    count, _ := strconv.Atoi(res[1])
    from, _ := strconv.Atoi(res[2])
    to, _ := strconv.Atoi(res[3])
    move := Move {
      count: count,
      from: from - 1,
      to: to - 1,
    }
    moves = append(moves, move)
  }

  return stacks, moves
}

func part1(stacks []string, moves []Move) string {
  for _, move := range(moves) {
    move.apply(stacks)
  }
  s := ""
  for _, stack := range(stacks) {
    s = fmt.Sprintf("%s%c", s, stack[len(stack)-1])
  }
  return s
}

func part2(stacks []string, moves []Move) string {
  for _, move := range(moves) {
    move.apply2(stacks)
  }
  s := ""
  for _, stack := range(stacks) {
    s = fmt.Sprintf("%s%c", s, stack[len(stack)-1])
  }
  return s
}

func main() {
  input := readInput()

  stacks, moves := parseStacks(input)
  fmt.Println(part1(stacks, moves))

  stacks, moves = parseStacks(input)
  fmt.Println(part2(stacks, moves))
}
